data_structures/bloom_filter.cpp

```diff
@@ -0,0 +1,292 @@
+ * @param size initial size of Bloom filter
+ * @param funks hash functions for T type
+ */
+template <typename T>
+BloomFilter<T>::BloomFilter(
+    std::size_t size,
+    std::initializer_list<std::function<std::size_t(T)>> funks)
+    : set(size), hashFunks(funks) {}
+
+/**
+ * @brief Add function for Bloom filter
+ *
+ * @tparam T type of elements that we need to filter
+ * @param x element to add to filter
+ * @return void
+ */
+template <typename T>
+void BloomFilter<T>::add(T x) {
+    for (std::size_t i = 0; i < hashFunks.size(); i++) {
+        set.add(hashFunks[i](x) % (sizeof(std::size_t) * set.size()));
+    }
+}
+
+/**
+ * @brief Check element function for Bloom filter
+ *
+ * @tparam T type of elements that we need to filter
+ * @param x element to check in filter
+ * @return true if the element probably appears in the filter
+ * @return false if the element certainly does not appear in the filter
+ */
+template <typename T>
+bool BloomFilter<T>::contains(T x) {
+    for (std::size_t i = 0; i < hashFunks.size(); i++) {
+        if (set.contains(hashFunks[i](x) %
+                         (sizeof(std::size_t) * set.size())) == false) {
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * @brief [Function djb2](http://www.cse.yorku.ca/~oz/hash.html), to get hash
+ * for string
+ *
+ * @param s string to get hash from
+ * @return hash for a string
+ */
+static std::size_t hashDJB2(std::string const& s) {
+    std::size_t hash = 5381;
+    for (char c : s) {
+        hash = ((hash << 5) + hash) + c;
+    }
+    return hash;
+}
+
+/**
+ * @brief [Hash
+ * function](https://stackoverflow.com/questions/8317508/hash-function-for-a-string),
+ * to get hash for string
+ *
+ * @param s string to get hash from
+ * @return hash for a string
+ */
+static std::size_t hashStr(std::string const& s) {
+    std::size_t hash = 37;
+    std::size_t primeNum1 = 54059;
+    std::size_t primeNum2 = 76963;
+    for (char c : s) {
+        hash = (hash * primeNum1) ^ (c * primeNum2);
+    }
+    return hash;
+}
+
+/**
+ * @brief [Hash function for
+ * test](https://stackoverflow.com/questions/664014/what-integer-hash-function-are-good-that-accepts-an-integer-hash-key)
+ *
+ * @param x int to get hash from
+ * @return hash for x
+ */
+std::size_t hashInt_1(int x) {
+    x = ((x >> 16) ^ x) * 0x45d9f3b;
+    x = ((x >> 16) ^ x) * 0x45d9f3b;
+    x = (x >> 16) ^ x;
+    return x;
+}
+
+/**
+ * @brief [Hash function for
+ * test](https://stackoverflow.com/questions/664014/what-integer-hash-function-are-good-that-accepts-an-integer-hash-key)
+ *
+ * @param x int to get hash from
+ * @return hash for x
+ */
+std::size_t hashInt_2(int x) {
+    auto y = static_cast<std::size_t>(x);
+    y = (y ^ (y >> 30)) * static_cast<std::size_t>(0xbf58476d1ce4e5b9);
+    y = (y ^ (y >> 27)) * static_cast<std::size_t>(0x94d049bb133111eb);
+    y = y ^ (y >> 31);
+    return y;
+}
+} // namespace data_structures
```

Comment:
```suggestion
}  // namespace data_structures
```


__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Tags:** Inefficient